File: sdks/python/pfp_sdk/canonicalize.py
"""PFP canonicalization — faithful port of the server's deterministic JSON rules.

Rules (must match backend/crypto/canonicalize.py byte-for-byte):
  1. Objects: keys sorted lexicographically; null values removed at every depth.
  2. Numbers: floats equal to a whole number collapse to int (100.0 -> 100).
  3. Strings matching ^\\d{4}-\\d{2}-\\d{2}T are normalized to
     YYYY-MM-DDTHH:MM:SS.mmmZ (millisecond precision, UTC).
  4. Arrays: each non-null element normalized, nulls dropped.
  5. Serialization: minimal separators (",", ":"), UTF-8, ensure_ascii=False.
"""
import json
import re
from datetime import datetime

_TS_RE = re.compile(r"\d{4}-\d{2}-\d{2}T")


def normalize_timestamp(ts: str) -> str:
    try:
        ts = ts.replace("Z", "+00:00")
        if "+" in ts or "-" in ts[10:]:
            dt = datetime.fromisoformat(ts)
        else:
            dt = datetime.fromisoformat(ts + "+00:00")
        return dt.strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"
    except Exception:
        return ts


def _normalize_value(value):
    if value is None:
        return None
    if isinstance(value, bool):
        return value
    if isinstance(value, int):
        return value
    if isinstance(value, float):
        if value == int(value):
            return int(value)
        return value
    if isinstance(value, str):
        if _TS_RE.match(value):
            return normalize_timestamp(value)
        return value
    if isinstance(value, list):
        return [_normalize_value(v) for v in value if v is not None]
    if isinstance(value, dict):
        return _canonicalize_dict(value)
    return value


def _canonicalize_dict(data: dict) -> dict:
    result = {}
    for key in sorted(data.keys()):
        value = data[key]
        if value is None:
            continue
        normalized = _normalize_value(value)
        if normalized is not None:
            result[key] = normalized
    return result


def canonicalize_to_json(data: dict) -> str:
    canonical = _canonicalize_dict(data)
    return json.dumps(canonical, separators=(",", ":"), ensure_ascii=False, sort_keys=True)

File: sdks/javascript/canonicalize.js
'use strict';
/**
 * PFP canonicalization — faithful port of backend/crypto/canonicalize.py.
 * Deterministic JSON: sorted keys, null-strip at all depths, float->int when
 * whole, ISO timestamp normalization to millisecond UTC, minimal separators.
 */
const TS_RE = /^\d{4}-\d{2}-\d{2}T/;

function normalizeTimestamp(ts) {
  const d = new Date(ts);
  if (isNaN(d.getTime())) return ts;
  return d.toISOString(); // YYYY-MM-DDTHH:MM:SS.mmmZ
}

function normalizeValue(value) {
  if (value === null || value === undefined) return undefined;
  if (typeof value === 'boolean') return value;
  if (typeof value === 'number') {
    return Number.isInteger(value) ? value : value;
  }
  if (typeof value === 'string') {
    return TS_RE.test(value) ? normalizeTimestamp(value) : value;
  }
  if (Array.isArray(value)) {
    return value.filter((v) => v !== null && v !== undefined).map(normalizeValue);
  }
  if (typeof value === 'object') return canonicalizeObject(value);
  return value;
}

function canonicalizeObject(obj) {
  const out = {};
  for (const key of Object.keys(obj).sort()) {
    const v = obj[key];
    if (v === null || v === undefined) continue;
    const nv = normalizeValue(v);
    if (nv !== undefined) out[key] = nv;
  }
  return out;
}

/** Serialize an already-canonical object with sorted keys + no whitespace. */
function stableStringify(value) {
  if (value === null) return 'null';
  if (Array.isArray(value)) return '[' + value.map(stableStringify).join(',') + ']';
  if (typeof value === 'object') {
    const parts = Object.keys(value)
      .sort()
      .map((k) => JSON.stringify(k) + ':' + stableStringify(value[k]));
    return '{' + parts.join(',') + '}';
  }
  return JSON.stringify(value);
}

function canonicalizeToJson(data) {
  return stableStringify(canonicalizeObject(data));
}

module.exports = { canonicalizeToJson, normalizeTimestamp };
